k8s/src/main/scala/io/buoyant/k8s/EventLogging.scala

```diff
@@ -1,55 +1,95 @@
 package io.buoyant.k8s
 
+import com.twitter.logging.Logger
+
 private[k8s] trait EventLogging {
   def ns: String
   def srv: String
+  
+  @inline 
+  protected def logAction[A](
+    noun: String,
+    format: A => String = (a: A) => a.toString
+  )(
+    verb: String
+  )(
+    value: A
+  ): Unit =
+    log.trace(
+      s"k8s ns %s service %s %s %s %s",
+      ns, srv, 
+      verb, noun,
+      format(value)
+    )
 
-  @inline private[this] def logAction[A](verb: String, descriptor: => String, format: A => String)(value: A): Unit =
-    log.ifTrace(s"k8s ns $ns service $srv $verb $descriptor ${format(value)}")
-
-  protected def logActions[A](verb: String, descriptor: => String, format: A => String)(values: Iterable[A]): Unit =
-    if (values.nonEmpty) {
-      log.debug(
-        "k8s ns %s service %s %s %ss",
-        ns, srv, verb, descriptor
-      )
-      values.foreach(logAction(verb, descriptor, format))
+  @inline
+  private[this] def formatMapping(kv: (Any, Any)): String =
+    kv match {
+      case (name: String, port: Int) => s"'$name' to port $port"
+      case (name: String, (oldTo, newTo)) => s"'$name' from $oldTo to $newTo"
+      case (name: String, to) => s"'$name' to $to"
+      case (fromPort, (oldTo, newTo)) => s"$fromPort from $oldTo to $newTo"
+      case (fromPort, to) => s"from $fromPort to $to"
     }
 
-  private[this] def formatMapping[A, B](kv: (A, B)): String = kv match {
-    case (name: String, port: Int) => s"'$name' to port $port"
-    case (name: String, to) => s"'$name' to $to"
-    case (fromPort, to) => s"from $fromPort to $to"
-  }
-
-  @inline def deletion[A, B](deleted: Map[A, B]): Unit =
-    logActions("deleted", "port mapping", formatMapping)(deleted)
-
-  @inline def addition[A, B](added: Map[A, B]): Unit =
-    logActions("added", "port mapping", formatMapping)(added)
-
-  protected def logModification[A](descriptor: String)(mods: Iterable[(A, A)]): Unit =
-    if (mods.nonEmpty) {
+  protected def _newState[A](
+    noun: String,
+    wasAdded: A => Unit,
+    wasDeleted: A => Unit
+  )(
+    oldState: Set[A],
+    newState: Set[A]
+  ): Unit =
+    if (oldState != newState) {
       log.debug(
         "k8s ns %s service %s modified %ss",
-        ns, srv, descriptor
+        ns, srv, noun
       )
-      for {
-        (old, replacement) <- mods
-      } logAction[A](s"replaced $old with", descriptor, _.toString)(replacement)
+      if (log.isLoggable(Logger.TRACE)) {
+        (newState -- oldState).foreach { wasAdded }
+        (oldState -- newState).foreach { wasDeleted }
+      }
     }
 
-  def modification[A, B](old: Map[A, B], replacement: Map[A, B]): Unit =
-    if (old.nonEmpty && replacement.nonEmpty) {
+  def deletion(noun: String = ""): Unit =
+    log.debug("k8s ns %s service %s deleted %s", ns, srv, noun)
+
+  private val mappingWas = logAction("port mapping", formatMapping)(_)
```

Comment:
it looks like this class is supposed to be type agnostic but this assumes that all mappings are port mappings?